README.md
# PostCSS gradientfixer [![Build Status][ci-img]][ci]

[PostCSS] plugin to rewrite code with existing vendor prefixes (especially -webkit-) on CSS gradients and add equivalent un-prefixed CSS.

[PostCSS]: https://github.com/postcss/postcss
[ci-img]:  https://travis-ci.org/hallvors/postcss-gradientfixer.svg
[ci]:      https://travis-ci.org/hallvors/postcss-gradientfixer

For example, here's some CSS with vendor-specific prefix and a gradient syntax based on older versions of the CSS specifications:

```css
.foo {
  background: -webkit-gradient(linear, left top, left bottom, color-stop(0, #ffffff), to(#f0efe7));
}
```

The unprefixer plugin will add the equivalent standard declarations (while leaving the old CSS for backwards compatibility with older WebKit-based browsers):

```css
.foo {
  background: -webkit-gradient(linear, left top, left bottom, color-stop(0, #ffffff), to(#f0efe7));
  background: linear-gradient(to bottom, #ffffff 0%, #f0efe7 100%);
}
```

## Usage

```js
postcss([ require('postcss-gradientfixer') ])
```

See [PostCSS] docs for examples for your environment.


CHANGELOG.md
# Gradient-fixer

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### Changed

## [0.0.6] - 2016-05-12
### Changed
- More tests and various fixes by gucong3000 on GitHub merged

## [0.0.5] - 2016-01-11
### Changed
- Update to postcss 5.x, avoid deprecated eachDecl function

## [0.0.4] - 2015-10-28
### Changed
- Tightens up the regular expressions for testing if there's existing gradient code in rule set

## [0.0.2] - 2015-06-06
### Changed
- More carefully checks if there's existing gradient code in rule set

## [0.0.0] - 2015-06-06
### Changed
- First release

test/test.js
var postcss = require('postcss');
var expect  = require('chai').expect;

var plugin = require('../');

var test = function (input, output, opts, done) {
    postcss([ plugin(opts) ]).process(input).then(function (result) {
        expect(result.css).to.eql(output);
        expect(result.warnings()).to.be.empty;
        done();
    }).catch(function (error) {
        done(error);
    });
};

describe('postcss-gradientfixer', function () {
    it('rewriting linear gradient', function (done) {
        test('b{background: -webkit-gradient(linear, left top, left bottom, color-stop(0, #ffffff), to(#f0efe7));}', 'b{background: -webkit-gradient(linear, left top, left bottom, color-stop(0, #ffffff), to(#f0efe7));background: linear-gradient(to bottom, #ffffff 0%, #f0efe7 100%);}', {}, done);
    });
    it('rewriting radial gradient', function (done) {
        test('b{background: -webkit-gradient(radial, 45 45, 10, 52 50, 30, from(#A7D30C), to(rgba(1,159,98,0)), color-stop(90%, #019F62)), -webkit-gradient(radial, 105 105, 20, 112 120, 50, from(#ff5f98), to(rgba(255,1,136,0)), color-stop(75%, #ff0188)), -webkit-gradient(radial, 95 15, 15, 102 20, 40, from(#00c9ff), to(rgba(0,201,255,0)), color-stop(80%, #00b5e2)), -webkit-gradient(radial, 0 150, 50, 0 140, 90, from(#f4f201), to(rgba(228, 199,0,0)), color-stop(80%, #e4c700));}', 'b{background: -webkit-gradient(radial, 45 45, 10, 52 50, 30, from(#A7D30C), to(rgba(1,159,98,0)), color-stop(90%, #019F62)), -webkit-gradient(radial, 105 105, 20, 112 120, 50, from(#ff5f98), to(rgba(255,1,136,0)), color-stop(75%, #ff0188)), -webkit-gradient(radial, 95 15, 15, 102 20, 40, from(#00c9ff), to(rgba(0,201,255,0)), color-stop(80%, #00b5e2)), -webkit-gradient(radial, 0 150, 50, 0 140, 90, from(#f4f201), to(rgba(228, 199,0,0)), color-stop(80%, #e4c700));background: radial-gradient(circle 30px at 45px 45px, #A7D30C 0%, #019F62 90%, rgba(1, 159, 98, 0) 100%), radial-gradient(circle 50px at 105px 105px, #ff5f98 0%, #ff0188 75%, rgba(255, 1, 136, 0) 100%), radial-gradient(circle 40px at 95px 15px, #00c9ff 0%, #00b5e2 80%, rgba(0, 201, 255, 0) 100%), radial-gradient(circle 90px at 0px 150px, #f4f201 0%, #e4c700 80%, rgba(228, 199, 0, 0) 100%);}', {}, done);
    });
    it('rewriting linear gradient with head', function (done) {
        test('b{background: url("logo.png"), -webkit-gradient(linear, left top, left bottom, from(#fff), to(#000));}', 'b{background: url("logo.png"), -webkit-gradient(linear, left top, left bottom, from(#fff), to(#000));background: url("logo.png"), linear-gradient(to bottom, #fff 0%, #000 100%);}', {}, done);
    });
    it('rewriting color-stop with out position', function (done) {
        test('b{background: -webkit-gradient(linear, left top, right bottom, from(black), color-stop(rgba(0, 0, 0, 0.5)), to(white));}', 'b{background: -webkit-gradient(linear, left top, right bottom, from(black), color-stop(rgba(0, 0, 0, 0.5)), to(white));background: linear-gradient(135deg, black 0%, rgba(0, 0, 0, 0.5) 50%, white 100%);}', {}, done);
    });
    it('rewriting 2 gradient in background-image', function (done) {
        test('b{background-image:-webkit-gradient( linear, left top, left bottom, from(white), to(black) ), -webkit-gradient( linear, left top, left bottom, from(black), to(white) );}', 'b{background-image:-webkit-gradient( linear, left top, left bottom, from(white), to(black) ), -webkit-gradient( linear, left top, left bottom, from(black), to(white) );background-image:linear-gradient(to bottom, white 0%, black 100%), linear-gradient(to bottom, black 0%, white 100%);}', {}, done);
    });
    it('rewriting list-style-image', function (done) {
        test('b{list-style-image: -webkit-gradient(linear, left top, left bottom, from(white), to(black));}', 'b{list-style-image: -webkit-gradient(linear, left top, left bottom, from(white), to(black));list-style-image: linear-gradient(to bottom, white 0%, black 100%);}', {}, done);
    });
    it('rewriting -webkit-mask', function (done) {
        test('b{-webkit-mask: -webkit-gradient(linear, left top, left bottom, from(white), to(black));}', 'b{-webkit-mask: -webkit-gradient(linear, left top, left bottom, from(white), to(black));mask: linear-gradient(to bottom, white 0%, black 100%);}', {}, done);
    });
    it('rewriting -webkit-linear-gradient', function (done) {
        test('b{background-image:-webkit-linear-gradient(top, #444444, #999999);}', 'b{background-image:-webkit-linear-gradient(top, #444444, #999999);background-image:linear-gradient(to bottom, #444444, #999999);}', {}, done);
    });
    it('rewriting -moz-linear-gradient', function (done) {
        test('b{background-image:-moz-linear-gradient(top, #444444, #999999);}', 'b{background-image:-moz-linear-gradient(top, #444444, #999999);background-image:linear-gradient(to bottom, #444444, #999999);}', {}, done);
    });
    it('rewriting -o-linear-gradient', function (done) {
        test('b{background-image:-o-linear-gradient(top, #444444, #999999);}', 'b{background-image:-o-linear-gradient(top, #444444, #999999);background-image:linear-gradient(to bottom, #444444, #999999);}', {}, done);
    });
});


index.js
var postcss = require( 'postcss' );

module.exports = postcss.plugin( 'postcss-flexboxfixer', function( opts ) {
    opts = opts || {};

    function getValueForProperty(parent, name, prefixAgnostic){
        var retValue;
        parent.walkDecls(prefixAgnostic ? new RegExp('^(?:-\\w+-)?' + name + '$') : name, function(decl) {
            retValue = decl.value;
        });
        return retValue;
    }

    function colorValue(obj){
        var ar = [];
        for (var i = 0; i < obj.length; i++) {
            ar.push(obj[i].name);
        }
        return ar.join(', ');
    }

    /* Given an array of args for "-webkit-gradient(...)" returned by
     * oldGradientParser(), this function constructs a string representing the
     * equivalent arguments for a standard "linear-gradient(...)" or
     * "radial-gradient(...)" expression.
     *
     * @param type  Either 'linear' or 'radial'.
     * @param args  An array of args for a "-webkit-gradient(...)" expression,
     *              provided by oldGradientParser() (not including gradient type).
     */
    function standardizeOldGradientArgs(type, args){
        var stdArgStr = '';
        var stops = [];
        if(/^linear/.test(type)){
            // linear gradient, args 1 and 2 tend to be start/end keywords
            var points = [].concat(args[0].name.split(/\s+/), args[1].name.split(/\s+/)); // example: [left, top, right, top]
            // Old webkit syntax "uses a two-point syntax that lets you explicitly state where a linear gradient starts and ends"
            // if start/end keywords are percentages, let's massage the values a little more..
            var rxPercTest = /\d+\%/;
            if(rxPercTest.test(points[0]) || points[0] === 0){
                var startX = parseInt(points[0]), startY = parseInt(points[1]), endX = parseInt(points[2]), endY = parseInt(points[3]);
                stdArgStr += Math.atan2(endY - startY, endX - startX) * (180 / Math.PI) + 90;
                stdArgStr += 'deg';
            }else{
                if(points[1] === points[3]){ // both 'top' or 'bottom, this linear gradient goes left-right
                    stdArgStr += 'to ' + points[2];
                }else if(points[0] === points[2]){ // both 'left' or 'right', this linear gradient goes top-bottom
                    stdArgStr += 'to ' + points[3];
                }else if(points[1] === 'top'){ // diagonal gradient - from top left to opposite corner is 135deg
                    stdArgStr += '135deg';
                }else{
                    stdArgStr += '45deg';
                }
            }

        }else if(/^radial/i.test(type)){ // oooh, radial gradients..
            stdArgStr += 'circle ' + args[3].name.replace(/(\d+)$/, '$1px') + ' at ' + args[0].name.replace(/(\d+) /, '$1px ').replace(/(\d+)$/, '$1px');
        }

        var toColor;
        var startStep = type === 'linear' ? 2 : 4;
        for (var j = startStep; j < args.length; j++) {
            var position, color, colorIndex;
            if(args[j].name === 'color-stop'){
                position = args[j].args[0].name;
                if (args[j].args[1]) {
                    colorIndex = 1;
                } else {
                    colorIndex = 0;
                    position = (j - startStep) / (args.length - startStep - 1);
                }
            }else if (args[j].name === 'to') {
                position = '100%';
                colorIndex = 0;
            }else if (args[j].name === 'from') {
                position = '0%';
                colorIndex = 0;
            }
            if (position >= 0 || position.indexOf('%') === -1) { // original Safari syntax had 0.5 equivalent to 50%
                position = parseFloat(position) * 100 + '%';
            }
            color = args[j].args[colorIndex].name;
            if (args[j].args[colorIndex].args) { // the color is itself a function call, like rgb()
                color += '(' + colorValue(args[j].args[colorIndex].args) + ')';
            }
            if (args[j].name === 'from'){
                stops.unshift(color + ' ' + position);
            }else if(args[j].name === 'to'){
                toColor = color;
            }else{
                stops.push(color + ' ' + position);
            }
        }

        // translating values to right syntax
        for(j = 0; j < stops.length; j++){
            stdArgStr += ', ' + stops[j];
        }
        if(toColor){
            stdArgStr += ', ' + toColor + ' 100%';
        }
        return stdArgStr;
    }

    function oldGradientParser(str){
        /** This method takes a legacy -webkit-gradient() method call and parses it
            to pull out the values, function names and their arguments.
            It returns something like [{name:'-webkit-gradient',args:[{name:'linear'}, {name:'top left'} ... ]}]
        */
        var objs = [{}], path = [], current, word = '', separatorChars = [',', '(', ')'];
        current = objs[0];
        path[0] = objs;
        //str = str.replace(/\s*\(/g, '('); // sorry, ws in front of ( would make parsing a lot harder
        for(var i = 0; i < str.length; i++){
            if(separatorChars.indexOf(str[i]) === -1){
                word += str[i];
            }else{ // now we have a "separator" - presumably we've also got a "word" or value
                current.name = word.trim();
                //GM_log(word+' '+path.length+' '+str[i])
                word = '';
                if(str[i] === '('){ // we assume the 'word' is a function, for example -webkit-gradient() or rgb(), so we create a place to record the arguments
                    if(!('args' in current)){
                        current.args = [];
                    }
                    current.args.push({});
                    path.push(current.args);
                    current = current.args[current.args.length - 1];
                    path.push(current);
                }else if(str[i] === ')'){ // function is ended, no more arguments - go back to appending details to the previous branch of the tree
                    current = path.pop(); // drop 'current'
                    current = path.pop(); // drop 'args' reference
                }else{
                    path.pop(); // remove 'current' object from path, we have no arguments to add
                    var currentParent = path[path.length - 1] || objs; // last object on current path refers to array that contained the previous "current"
                    currentParent.push({}); // we need a new object to hold this "word" or value
                    current = currentParent[currentParent.length - 1]; // that object is now the 'current'
                    path.push(current);
    //GM_log(path.length)
                }
            }
        }

        return objs;
    }
    function createFixupGradientDeclaration(prop, value){
        value = value.trim();
        var newValue = '', rxfix, i;
        prop = postcss.vendor.unprefixed(prop);
        // -webkit-gradient(<type>, <point> [, <radius>]?, <point> [, <radius>]? [, <stop>]*)
        // fff -webkit-gradient(linear,0 0,0 100%,from(#fff),to(#f6f6f6));
        // Sometimes there is code before the -webkit-gradient, for example when it's part of a more complex background: shorthand declaration
        // we'll extract and keep any stuff before -webkit-gradient before we try parsing the gradient part
        var head = value.substr(0, value.indexOf('-webkit-gradient'));
        if(head){
            value = value.substr(head.length);
        }
        var m = value.match(/-webkit-gradient\s*\(\s*(linear|radial)\s*(.*)/);
        if(m){ // yay, really old syntax...

            // extracting the values..
            var parts = oldGradientParser(value), type; //GM_log(JSON.stringify(parts, null, 2))
            for(i = 0; i < parts.length; i++){
                if(!parts[i].args)continue;
                if(parts[i].name === '-webkit-gradient'){
                    type = parts[i].args[0].name;
                    newValue += type + '-gradient('; // radial or linear
                    rxfix = new RegExp('(^|\s)' + type + '-gradient\\(', '');
                }
                newValue += standardizeOldGradientArgs(type, parts[i].args.slice(1));
                newValue += ')'; // end of gradient method
                if (i < parts.length - 1) {
                    newValue += ', ';
                }
            }
        }else{ // we're dealing with more modern syntax - should be somewhat easier, at least for linear gradients.
            // Fix three things: remove -webkit-, add 'to ' before reversed top/bottom keywords (linear) or 'at ' before position keywords (radial), recalculate deg-values
            // -webkit-linear-gradient( [ [ <angle> | [top | bottom] || [left | right] ],]? <color-stop>[, <color-stop>]+);
            newValue = value.replace(/(^|\s|,)-\w+-/g, '$1');
            // Keywords top, bottom, left, right: can be stand-alone or combined pairwise but in any order ('top left' or 'left top')
            // These give the starting edge or corner in the -webkit syntax. The standardised equivalent is 'to ' plus opposite values for linear gradients, 'at ' plus same values for radial gradients
            if(newValue.indexOf('linear') > -1){
                newValue = newValue.replace(/(top|bottom|left|right)+\s*(top|bottom|left|right)*/, function(str){
                    var words = str.split(/\s+/);
                    for(i = 0; i < words.length; i++){
                        switch(words[i].toLowerCase()){
                            case 'top':
                                words[i] = 'bottom';
                                break;
                            case 'bottom':
                                words[i] = 'top';
                                break;
                            case 'left':
                                words[i] = 'right';
                                break;
                            case 'right':
                                words[i] = 'left';
                        }
                    }
                    str = words.join(' ');
                    return 'to ' + str;
                });
                rxfix = /(^|\s)linear-gradient\(/;
            }else{
                newValue = newValue.replace(/(top|bottom|left|right)+\s/, 'at $1 ');
                rxfix = /(^|\s)radial-gradient\(/;
            }

            newValue = newValue.replace(/\d+deg/, function (val) {
                 return 360 - parseInt(val) - 90 + 'deg';
             });

        }
        // Similar to the code for "head" above..
        var tail = value.substr(value.lastIndexOf(')') + 1);
        if( tail && tail.trim() !== ','){ // sometimes the gradient is a part of a more complex declaration (for example an image shorthand with stuff like no-repeat added), and what's after the gradient needs to be included
            newValue += tail;
        }
        if(head){
            newValue = head + newValue;
        }
        // GM_log(newValue)
        return {type:'declaration', property:prop, value:newValue, rxfix:rxfix};
    }

    return function( css ) {
        css.walkDecls(function(decl) {
            if(!/-(?:webkit|\w+(?:-\w+)+)-gradient/.test(decl.value)){
                /* no -webkit- gradient syntax here, move on */
                return;
            }
            var fixedDecl = createFixupGradientDeclaration(decl.prop, decl.value);
            /* we only add fixup rules if there's no equivalent rule in the CSS rule set already.. */
            var existingValue = getValueForProperty(decl.parent, decl.prop, false);
            if(!(existingValue === fixedDecl.value || fixedDecl.rxfix.test(existingValue))){
                decl.cloneAfter({'prop':fixedDecl.property, 'value':fixedDecl.value});
            }
        } );
    };
} );
